<a href="https://www.kaggle.com/code/charleschuang/python-class?scriptVersionId=134319256" target="_blank"><img align="left" alt="Kaggle" title="Open in Kaggle" src="https://kaggle.com/static/images/open-in-kaggle.svg"></a>

# Python 類別(class)實作範例

## Step 1 為了瞭解如何實作class與物件屬性，進行簡單範例練習。

> 假如我們認識一個人，最快的方式是從外貌及基本資訊來認識。例如: 年齡、體重、身高、出生年月日、性別、死活狀態等。

> 為了往後可以將設計好的框架套用在不同的人，python可以讓我們事前先設計一個類別框架。因此，這裡宣告一個類別(Class)叫做Person，藉此涵蓋未來要收納的各種不同的人物，可以把class想像成一本神奇寶貝收錄圖鑑，小智起初只會有空白的圖鑑，隨著收錄的神奇寶貝越多，則可以用圖鑑呼叫出個別神奇寶貝的相關資訊。

> 空白的圖鑑需要設計基本的框架，才能作為後續的套用。套用方法，可以先試想，我們今天要將一個人的相關資訊輸入至變數，最方便的就是用功能(function)，搭配對應的參數，記錄在該變數名稱當中。當我需要查詢時，只要再輸入該變數名稱及對應的參數，就能獲得資訊。假想的目標如下:

In [66]:
charles = Person(18, 60, 170, 20230525,"male") #先記錄在charles這個變數
charles.age #查詢年齡，會返回18歲

1018

> 為了達成該目的，下列開始進行如何撰寫程式碼。基本上，要用 __init__來初始化，也就是先設計神奇寶貝圖鑑應該要有哪些欄位資訊。def則是函式、功能、方法(function, method)宣告，def後方設定為__init__，該方法則被稱為class所宣告的: Person，而非__init__。內層當中皆以self.[attribute/featrue]來設計特徵，self指的就是未來變數，例如:charles。

> 接著為了能夠回傳其存活狀態的資訊，另外以def宣告livestatus()這個功能，如果存活則回傳"lived"，相反的，如果已死亡，則回傳"dead"。如此一來，則成功加入__init__以外的功能在該類別(class)當中。

In [65]:
class Person: #宣告類別: 宣告該類別為person
    def __init__(self, age, weight, height, birth, gender,live = 1): #宣告方法 (method or function)
        self.age = age+1000 #宣告屬性 年齡
        self.weight = weight #宣告屬性 體重
        self.height = height #宣告屬性 身高
        self.birth = birth #宣告屬性: 出生年月日
        self.gender = gender #宣告屬性: 性別
        self.live = live #宣告屬性: 死活
    def livestatus(self): #宣告方法 or 功能函式
        if self.live == 1:
            return "lived"
        else: 
            return "dead"  
    class chilgren:  #class 裡面可以再塞class
        def __init__(self, number):
            self.number = number
        def sayhello(self): #class 裡面可以再塞function
            return "Hello, World!"

## Step 2 類別繼承
> 此方法只繼承屬性與方法的框架，無法連通父系內部更細微的操作，除非使用Super

In [124]:
class Superman(Person):
    def __init__(self,age,superpower):
        self.superpower = superpower
        self.age = age
        print("This superman has the power:"+ superpower)

> 此時不會要求你一定要輸入父系的屬性值，因此父系屬性值目前皆為空值。然而，重新定義的age必須輸入，但只會根據這裡重新定義的方式回傳結果，與父系無關。

In [125]:
S = Superman(18,"橡膠果實")

This superman has the power:橡膠果實


In [126]:
S.age

18

> 添加值到那些從父系類別繼承的屬性當中

In [127]:
S.age =18
S.birth = 19930722
S.weight = 60

In [128]:
S.age , S.birth, S.weight , S.superpower

(18, 19930722, 60, '橡膠果實')

## Step 3 Super 用法

https://iampennywu.gitbooks.io/introducing-python/content/chapter6-5.html
當父系類別的屬性和方法變動時，繼承的子類別同樣能反映並自動更新，就要用到super

In [129]:
def __init__(self, name, email):
            super().__init__(name)         # super() 取得父類別的定義

> 此時將要求起始(init)階段即填入父系屬性的值

In [130]:
class Superman_with_super(Person):
    def __init__(self,age, weight, height, birth, gender,live, superpower):
        super().__init__(age, weight, height, birth, gender,live)
        self.superpower = superpower
        print("This superman has the power:"+ superpower)

> 此時輸入值則帶入至父系內部的細微操作，例如: age+1000。而非僅有繼承其框架而已

In [131]:
S = Superman_with_super(18, 60, 170, 20230525,"male",1,"橡膠果實")

This superman has the power:橡膠果實


> 將class person 的age 調整成+1000，Superman_with_super可以直接連通，輸入18可以回傳1018

In [132]:
S.age , S.birth , S.superpower

(1018, 20230525, '橡膠果實')

>　若另外賦予新值給age，則跳脫父系的 +1000設定，直接回傳18

In [133]:
S.age = 18

In [134]:
S.age

18